File: backend/src/outreach/schemas/ai_sequences.py
from __future__ import annotations

from datetime import time
from typing import Literal

from pydantic import BaseModel, Field

from outreach.schemas.sequences import SequenceDetail, StepOut


class KnowledgeUploadResult(BaseModel):
    knowledge_id: str
    files_processed: int
    chunks_indexed: int
    filenames: list[str]


class AISequenceGenerateRequest(BaseModel):
    prompt: str = Field(min_length=10, max_length=8000)
    knowledge_id: str | None = Field(default=None, max_length=64)


class AIStepDraft(BaseModel):
    channel: Literal[
        "email", "linkedin_dm", "linkedin_connect", "linkedin_like", "call", "sms", "whatsapp"
    ]
    delay_days: int = Field(default=0, ge=0, le=365)
    delay_hours: int = Field(default=0, ge=0, le=23)
    subject: str | None = None
    body: str
    config: dict = Field(default_factory=dict)


class AISequenceDraft(BaseModel):
    name: str
    description: str | None = None
    timezone: str = "Asia/Kolkata"
    send_window_start: str = "09:00"
    send_window_end: str = "18:00"
    send_days_mask: int = 31
    ai_followups_enabled: bool = True
    steps: list[AIStepDraft]


class AISequenceGenerateResponse(BaseModel):
    draft: AISequenceDraft
    knowledge_snippets_used: int


class AISequenceSaveRequest(BaseModel):
    draft: AISequenceDraft
    knowledge_id: str | None = None


class AIStepRegenerateRequest(BaseModel):
    prompt: str | None = Field(default=None, max_length=4000)


class AIStepRegenerateResponse(BaseModel):
    step: StepOut
    sequence: SequenceDetail

File: frontend/src/api/aiSequences.ts
import { api } from "./client";
import type { SequenceDetail, StepOut } from "./sequences";

export type AIStepChannel =
  | "email"
  | "linkedin_dm"
  | "linkedin_connect"
  | "linkedin_like"
  | "call"
  | "sms"
  | "whatsapp";

export interface KnowledgeUploadResult {
  knowledge_id: string;
  files_processed: number;
  chunks_indexed: number;
  filenames: string[];
}

export interface AIStepDraft {
  channel: AIStepChannel;
  delay_days: number;
  delay_hours: number;
  subject: string | null;
  body: string;
  config: Record<string, unknown>;
}

export interface AISequenceDraft {
  name: string;
  description: string | null;
  timezone: string;
  send_window_start: string;
  send_window_end: string;
  send_days_mask: number;
  ai_followups_enabled: boolean;
  steps: AIStepDraft[];
}

export interface AISequenceGenerateResponse {
  draft: AISequenceDraft;
  knowledge_snippets_used: number;
}

export async function uploadKnowledge(files: File[]): Promise<KnowledgeUploadResult> {
  const form = new FormData();
  files.forEach(f => form.append("files", f));
  return (await api.post<KnowledgeUploadResult>("/ai-sequences/knowledge/upload", form, {
    headers: { "Content-Type": "multipart/form-data" },
  })).data;
}

export async function generateAISequence(prompt: string, knowledgeId?: string | null) {
  return (await api.post<AISequenceGenerateResponse>(
    "/ai-sequences/generate",
    { prompt, knowledge_id: knowledgeId || null },
    { timeout: 120_000 },
  )).data;
}

export async function saveAISequence(draft: AISequenceDraft, knowledgeId?: string | null) {
  return (await api.post<SequenceDetail>("/ai-sequences/save", {
    draft,
    knowledge_id: knowledgeId || null,
  })).data;
}

export async function regenerateAIStep(
  sequenceId: number,
  stepId: number,
  prompt?: string | null,
) {
  return (await api.post<{ step: StepOut; sequence: SequenceDetail }>(
    `/ai-sequences/sequences/${sequenceId}/steps/${stepId}/regenerate`,
    { prompt: prompt || null },
  )).data;
}
